### PR TITLE
Databrowser jiras 84, 72, 93 [risk = no]

### PR DESCRIPTION
**Pull request by SrushtiGangireddy**
Addressing data browser jiras 84, 72, 93.
1. Db - 84 Update text content through out app
2. Db - 72 Update ui for code search results
3. Db - 93 Quick search ehr domain typing in 001 the snomed concept doesn't work initially.
4. Db - 83 add concept ids to data browser

### CODE REVIEW

File: public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java [hunk 1]
```diff
@@ -424,15 +427,25 @@ public ConceptAnalysis apply(ConceptAnalysis ca) {
                 if(!conceptSynonymNames.contains(conceptSynonym.getConceptSynonymName()) && !con.getConceptName().equals(conceptSynonym.getConceptSynonymName())){
                     conceptSynonymNames.add(conceptSynonym.getConceptSynonymName());
                     }
-            }
```
Comment (biopete) [inline public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java]: nit: remove this extra line

File: public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java [hunk 2]
```diff
@@ -424,15 +427,25 @@ public ConceptAnalysis apply(ConceptAnalysis ca) {
...
 
+            }
 
             this.conceptSynonymNames.put(con.getConceptId(),conceptSynonymNames);
 
             if((con.getStandardConcept() == null || !con.getStandardConcept().equals("S") ) && (searchConceptsRequest.getQuery().equals(conceptCode) || searchConceptsRequest.getQuery().equals(conceptId))){
                 response.setMatchType(conceptCode.equals(searchConceptsRequest.getQuery()) ? MatchType.CODE : MatchType.ID );
 
                 List<Concept> std_concepts = conceptDao.findStandardConcepts(con.getConceptId());
+                for(Concept concept: std_concepts){
```
Comment (biopete) [inline public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java]: I think you should update the conceptDao.findStandardConcepts to fetch the synonyms since we have all the relationship setup and everything and it's not a lot of rows. Then you don't need this code.

File: public-ui/src/app/views/ehr-view/ehr-view.component.html [hunk 3]
```diff
@@ -65,33 +65,61 @@ <h5 class="col-xs-12 box-title">
...
+              maps to Standard Vocabulary {{standardConcepts[0].vocabularyId}} {{standardConcepts[0].conceptCode}}.
+              Standard vocabularies capture data across a variety of source vocabularies.
             </clr-alert>
           </div>
         </div>
         <div class="row">
-          <div class="col-md grid-header flex-xs-middle"> {{dbDomain.domainDisplay}}</div>
-          <div class="col-md grid-header flex-xs-middle"> Synonyms </div>
-          <div class="col-md-2 grid-header flex-xs-middle"> Participants of <b> {{totalParticipants | number}}</b> </div>
+          <div class="col-md-3 grid-header flex-xs-middle"> Concept Name</div>
```
Comment (biopete) [inline public-ui/src/app/views/ehr-view/ehr-view.component.html]: Is there a reason not to use domainInfo.name here ?

Comment (biopete) [reply public-ui/src/app/views/ehr-view/ehr-view.component.html]: I kind of like how the header says Condition or whatever.

File: public-ui/src/app/views/ehr-view/ehr-view.component.html [hunk 4]
```diff
@@ -65,33 +65,61 @@ <h5 class="col-xs-12 box-title">
...
-
         </div>
         <div class="row expandable-grid-row"
-             *ngFor="let r of items.concat(standardConcepts), let i = index"
+             *ngFor="let r of standardConcepts"
              [ngClass]="{'expanded-grid-row': r.expanded}" >
-          <div class="col-md flex-xs-middle" innerHtml="{{i+1}}. {{r.conceptName | highlightSearch : prevSearchText}}
+          <div class="col-md-3 flex-xs-middle" innerHtml=" <b>Standard</b> {{r.conceptName | highlightSearch : prevSearchText}}
+            -- {{r.conceptId}}"> </div>
+          <div class="col-md-4 flex-xs-middle" *ngIf="r.conceptSynonyms;else noConceptSynonyms" innerHtml="{{ r.conceptSynonyms.join(', ') | highlightSearch : prevSearchText}}"></div>
```
Comment (biopete) [inline public-ui/src/app/views/ehr-view/ehr-view.component.html]: Why the ngIf ? Does this give an error if there are no synonyms ?  

Also could you add a custom pipe like this to truncate the synonyms to the length we want ? I think that will be a good safeguard to have in place.

Comment (SrushtiGangireddy) [reply public-ui/src/app/views/ehr-view/ehr-view.component.html]: Yeah, it was throwing error before on using join on undefined.

File: public-ui/src/app/views/ehr-view/ehr-view.component.html [hunk 5]
```diff
@@ -65,33 +65,61 @@ <h5 class="col-xs-12 box-title">
...
+              <clr-icon shape="bar-chart" style="height: 1rem; width: 1rem;color: #2691D0;"></clr-icon>
+            </button>
+          </div>
+          <div class="col-xs-12 row-expansion flex-xs-middle" *ngIf="r.expanded">
+            <app-concept-charts [concept]="r" [showMeasurementGenderBins]="showMeasurementGenderBins" [showAge]="showAge" [showGender]="showGender"
+                                [showSources]="showSources" [backgroundColor]="'#ECF1F4'">
+            </app-concept-charts>
+          </div>
+        </div>
+        <div class="row expandable-grid-row"
```
Comment (biopete) [inline public-ui/src/app/views/ehr-view/ehr-view.component.html]: I think we can still do these in just one loop instead of repeating this huge block of code. Instead of  *ngFor="let r of items.concat(standardConcepts), let i = index" 
do 
 *ngFor="let r of standardConcepts.concat(items), let i = index"

File: public-ui/src/app/views/ehr-view/ehr-view.component.html [hunk 6]
```diff
@@ -65,33 +65,61 @@ <h5 class="col-xs-12 box-title">
...
+          <div class="col-md-3 flex-xs-middle" *ngIf="matchType === 'CONCEPT_CODE' && r.standardConcept !='S';else not_code_match" innerHtml="<b>Source</b> {{r.conceptName | highlightSearch : prevSearchText}}
+            -- {{r.conceptId}}"> </div>
+          <ng-template #not_code_match>
+            <div class="col-md-3 flex-xs-middle" innerHtml="{{i+1}}. {{r.conceptName | highlightSearch : prevSearchText}}
             -- {{r.conceptId}}"> </div>
-          <div class="col-md flex-xs-middle" innerHtml="{{ r.conceptSynonyms.join(', ') | highlightSearch : prevSearchText}}"> </div>
-          <div class="col-md-2 flex-items-xs-right">
+          </ng-template>
+            <div class="col-md-4 flex-xs-middle" *ngIf="r.conceptSynonyms;else noConceptSynonyms" innerHtml="{{ r.conceptSynonyms.join(', ') | highlightSearch : prevSearchText}}"></div>
+            <ng-template #noConceptSynonyms><div class="col-md-2 flex-xs-middle"></div></ng-template>
```
Comment (biopete) [inline public-ui/src/app/views/ehr-view/ehr-view.component.html]: if you really need this ngIf stuff , need to make that a col-md-4 i think ..?

File: public-ui/src/app/views/ehr-view/ehr-view.component.html [hunk 7]
```diff
@@ -65,33 +65,61 @@ <h5 class="col-xs-12 box-title">
...
+          <div class="col-md-2 flex-items-xs-right" style="margin-top: .5rem;">
```
Comment (biopete) [inline public-ui/src/app/views/ehr-view/ehr-view.component.html]: No inline styles, please. Use the ehr-view.component.css

File: public-ui/src/app/views/ehr-view/ehr-view.component.ts [hunk 8]
```diff
@@ -126,6 +127,10 @@ export class EhrViewComponent implements OnInit, OnDestroy {
   private searchCallback(results: any) {
     this.searchResult = results;
     this.items = this.searchResult.items;
+    this.standardConcepts = this.searchResult.standardConcepts;
```
Comment (biopete) [inline public-ui/src/app/views/ehr-view/ehr-view.component.ts]: duplicate? Or does this return an empty array if there are none? We may not need the if's below.

File: public-ui/src/app/views/ehr-view/ehr-view.component.ts [hunk 9]
```diff
@@ -126,6 +127,10 @@ export class EhrViewComponent implements OnInit, OnDestroy {
...
+    if (this.searchResult.matchType) {
```
Comment (biopete) [inline public-ui/src/app/views/ehr-view/ehr-view.component.ts]: Is matchType sometimes not set? Could we just use searchResult.matchType in our template and not need the separate variable? For items and standard concepts i wanted them to be empty arrays and wasn't sure if the api would return that for each.

File: public-ui/src/app/views/quick-search/quick-search.component.css [hunk 10]
```diff
@@ -12,15 +12,15 @@
 
 /* result card */
 .white-card-box.search-result {
-    height: 240px;
-    width: 240px;
+    height: 260px;
```
Comment (biopete) [inline public-ui/src/app/views/quick-search/quick-search.component.css]: So for sizing I've been trying to use multiples of 24px. Clarity recommends that for the grid as that is what all elements are based on. See here: https://vmware.github.io/clarity/documentation/v0.12/app-layout

  So 264px may be better and 246px may be better for width.

File: public-ui/src/app/views/quick-search/quick-search.component.html [hunk 11]
```diff
@@ -38,12 +38,7 @@ <h5 class="box-title">
               {{r.countValue | number}}
             </div>
             <p class="card-text bigger">
-              <em> <b>concepts </b></em>
-              <span *ngIf="searchText.value"> match your search. </span>
```
Comment (biopete) [inline public-ui/src/app/views/quick-search/quick-search.component.html]: Why did we get rid of this  if for search or not?

File: public-ui/src/app/views/quick-search/quick-search.component.ts [hunk 12]
```diff
@@ -49,9 +49,13 @@ export class QuickSearchComponent implements OnInit, OnDestroy {
     // Set title based on datatype
     if (this.dataType === this.EHR_DATATYPE) {
       this.title = 'Electronic Health Data';
+      this.subTitle = 'Enter a keyword or data standards code (eg ICD, SNOMED) in the search bar ' +
+          'to search across Electronic Health Record (EHR) data and program surveys.';
```
Comment (biopete) [inline public-ui/src/app/views/quick-search/quick-search.component.ts]: Remove ehr surveys here

File: public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java (hunk 1 above)
Comment (SrushtiGangireddy) [reply public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java]: Done.

File: public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java (hunk 2 above)
Comment (SrushtiGangireddy) [reply public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java]: It was not possible to join fetch with concept relationship and concept synonym at the same time and the concept synonyms were not fetched by default as we specified it to be lazy. So added a query in dao to join fetch concept synonyms of only standard concept by id.

File: public-ui/src/app/views/ehr-view/ehr-view.component.ts (hunk 9 above)
Comment (SrushtiGangireddy) [reply public-ui/src/app/views/ehr-view/ehr-view.component.ts]: Yeah, changed it to use the searchResult.matchType.

File: public-ui/src/app/views/ehr-view/ehr-view.component.ts (hunk 8 above)
Comment (SrushtiGangireddy) [reply public-ui/src/app/views/ehr-view/ehr-view.component.ts]: I added else part to set standard concept to empty array because i saw it getting the previously fetched standard concepts as part of next search.

File: public-ui/src/app/views/quick-search/quick-search.component.css (hunk 10 above)
Comment (SrushtiGangireddy) [reply public-ui/src/app/views/quick-search/quick-search.component.css]: Cool thanks. That was helpful. I had to change this to fit in survey description.

File: public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java (hunk 2 above)
Comment (biopete) [reply public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java]: Hmm.. Ok. In that case i would use this is ConceptService ...
https://github.com/all-of-us/workbench/blob/master/common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptService.java#L230

File: common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java [hunk 13]
```diff
@@ -13,6 +13,10 @@
             "where cr.concept_id_1=?1 and cr.relationship_id='Maps to' ")
     List<Concept> findStandardConcepts(long concept_id);
 
+    @Query("select c from Concept c left join FETCH c.synonyms as cs " +
```
Comment (biopete) [inline common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java]: Check out function in ConceptService that does this .. See below.

File: public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java [hunk 14]
```diff
@@ -360,22 +358,32 @@ public ConceptAnalysis apply(ConceptAnalysis ca) {
...
 
                 List<Concept> std_concepts = conceptDao.findStandardConcepts(con.getConceptId());
-                response.setStandardConcepts(std_concepts.stream().map(TO_CLIENT_CONCEPT).collect(Collectors.toList()));
+                List<Concept> stdConceptswithSynonyms = new ArrayList<>();
+                for(Concept concept: std_concepts){
+                    Concept conceptWithSynonyms = conceptDao.fetchConceptWithSynonyms(concept.getConceptId());
+                    stdConceptswithSynonyms.add(conceptWithSynonyms);
+                    ArrayList<String> uniqueConceptSynonymNames = new ArrayList<>();
+                    for(ConceptSynonym cs:conceptWithSynonyms.getSynonyms()){
+                        if(!uniqueConceptSynonymNames.contains(cs.getConceptSynonymName()) && !concept.getConceptName().equals(cs.getConceptSynonymName())){
```
Comment (biopete) [inline public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java]: So you are just removing the concept name from synonm list right? There really aren't any duplicates ... We can ennforce that on the database level. 

So i would  do this here:  https://github.com/all-of-us/workbench/blob/master/common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptService.java#L230

That function i'm sure wants the duplicate removed.

File: public-ui/src/app/views/ehr-view/ehr-view.component.css [hunk 15]
```diff
@@ -7,3 +7,6 @@
 .white-card-box.top-stat {
     padding: 1rem;
 }
+.flex-items-xs-right {
```
Comment (biopete) [inline public-ui/src/app/views/ehr-view/ehr-view.component.css]: Please don't put custom styles on built classes used by grid system. make your own class and add it to it.  

These built in classes are meant to do a very particular thing and are not intended to customize unless you want that customization to happen everywhere on the page. Which in general we dont.

File: public-ui/src/app/views/ehr-view/ehr-view.component.html [hunk 16]
```diff
@@ -65,32 +65,41 @@ <h5 class="col-xs-12 box-title">
...
              [ngClass]="{'expanded-grid-row': r.expanded}" >
-          <div class="col-md flex-xs-middle" innerHtml="{{i+1}}. {{r.conceptName | highlightSearch : prevSearchText}}
+          <div class="col-md-3 flex-xs-middle" *ngIf="searchResult.matchType === 'CONCEPT_CODE' && r.standardConcept !='S';else code_match_standard"
+               innerHtml="<b>Source</b> {{r.conceptName | highlightSearch : prevSearchText}}
+            -- {{r.conceptId}}"> </div>
+          <ng-template #code_match_standard>
+            <div class="col-md-3 flex-xs-middle" *ngIf="searchResult.matchType === 'CONCEPT_CODE' && r.standardConcept ==='S';else code_match_not_standard" innerHtml="<b>Standard</b> {{r.conceptName | highlightSearch : prevSearchText}}
+            -- {{r.conceptId}}"> </div>
+          </ng-template>
+          <ng-template #code_match_not_standard>
```
Comment (biopete) [inline public-ui/src/app/views/ehr-view/ehr-view.component.html]: Is this used?

File: public-ui/src/app/views/ehr-view/ehr-view.component.ts [hunk 17]
```diff
@@ -128,6 +128,8 @@ export class EhrViewComponent implements OnInit, OnDestroy {
     this.items = this.searchResult.items;
     if (this.searchResult.standardConcepts) {
       this.standardConcepts = this.searchResult.standardConcepts;
+    } else {
```
Comment (biopete) [inline public-ui/src/app/views/ehr-view/ehr-view.component.ts]: You don't need this else because we initialize to empty array at the top.

File: public-ui/src/app/views/quick-search/quick-search.component.html [hunk 18]
```diff
@@ -35,15 +35,11 @@ <h5 class="box-title">
               {{r.name}}
             </h5>
             <div class="card-stat">
-              {{r.standardConceptCount | number}}
+              {{r.allConceptCount | number}}
```
Comment (biopete) [inline public-ui/src/app/views/quick-search/quick-search.component.html]: We want standardConceptCount here i would think ...

File: public-ui/src/app/views/quick-search/quick-search.component.html [hunk 19]
```diff
@@ -35,15 +35,11 @@ <h5 class="box-title">
...
             </div>
             <p class="card-text bigger">
-              <em> <b>concepts </b></em>
-              <span *ngIf="searchText.value"> match your search. </span>
-              <span *ngIf="!searchText.value">in this domain.</span>
-              <span *ngIf="r.participantCount">
-                <span class="bolder"> {{r.participantCount | number}}</span> participants in this domain
-              </span>
+              <span *ngIf="searchText.value; else no_search_value">matching medical concepts</span>
+              <ng-template #no_search_value><span>medical concepts</span></ng-template>
```
Comment (biopete) [inline public-ui/src/app/views/quick-search/quick-search.component.html]: I don't recall the big changes to this part being required. But if they are great.

File: public-ui/src/app/views/quick-search/quick-search.component.html [hunk 20]
```diff
@@ -68,10 +64,11 @@ <h5 class="box-title">
             {{r.questionCount | number}}
           </div>
           <p class="card-text bigger">
-            There are {{r.questionCount }} survey questions on this survey related to your search.
+             matching survey questions <br>
```
Comment (biopete) [inline public-ui/src/app/views/quick-search/quick-search.component.html]: Again, was this change required?  Was it part of Kayla's ticket?

File: public-ui/src/app/views/survey-view/survey-view.component.ts [hunk 21]
```diff
@@ -130,7 +129,7 @@ export class SurveyViewComponent implements OnInit, OnDestroy {
 
   public countPercentage(countValue: number) {
     if (!countValue || countValue <= 0) { return 0; }
-    let percent: number = countValue / this.survey.countValue ;
+    let percent: number = countValue / this.survey.participantCount ;
     percent = parseFloat(percent.toFixed(2));
```
Comment (biopete) [inline public-ui/src/app/views/survey-view/survey-view.component.ts]: Can you change this to 4 and see what happens ? Because we end up with a lot of 0% in the display for answers not answered a lot. Better to see 0.01 ... Just simply change that to 2 to a 4

File: public-ui/src/app/views/ehr-view/ehr-view.component.html (hunk 16 above)
Comment (SrushtiGangireddy) [reply public-ui/src/app/views/ehr-view/ehr-view.component.html]: yeah, this is used for displaying the number and concept name if the match type is not concept code.

File: public-ui/src/app/views/quick-search/quick-search.component.ts [hunk 22]
```diff
@@ -13,140 +13,144 @@ import {DomainInfosAndSurveyModulesResponse} from '../../../publicGenerated/mode
...
-  EHR_DATATYPE = 'ehr';
-  SURVEY_DATATYPE = 'surveys';
+    pageImage = '/assets/db-images/man-standing.png';
+    title = 'Quick Guided Search';
+    subTitle = 'Enter a keyword or data standards code (eg ICD, SNOMED) in the search bar ' +
+        'to search across Electronic Health Record (EHR) data and program surveys.';
+    searchResults = [];
+    domainResults = [];
+    surveyResults = [];
+    totalDomainResults = [];
```
Comment (biopete) [inline public-ui/src/app/views/quick-search/quick-search.component.ts]: Just save the one object DomainandSurveyResults here instead of two separate variables.

File: public-ui/src/app/views/quick-search/quick-search.component.ts [hunk 23]
```diff
@@ -13,140 +13,144 @@ import {DomainInfosAndSurveyModulesResponse} from '../../../publicGenerated/mode
...
+        this.loading = false;
+    }
+    public searchDomains(query: string) {
+        this.prevSearchText = query;
+        localStorage.setItem('searchText', query);
+        // If query empty reset to already retrieved domain totals
+        if (query.length === 0) {
+            const resultsObservable = new Observable((observer) => {
+                const domains: DomainInfosAndSurveyModulesResponse = {
+                    domainInfos: this.totalDomainResults,
```
Comment (biopete) [inline public-ui/src/app/views/quick-search/quick-search.component.ts]: use the object.DomainInfos and object.SurveyModules ..

File: api/db-cdr/changelog-schema/db.changelog-18.xml [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+<databaseChangeLog
+    xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+    xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.0.xsd
+        http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd">
+  <changeSet author="danrodney" id="changelog-18">
+
+    <addColumn tableName="concept">
+      <!-- Contains the concept ID, concept code, concept name, and all concept_synonym_name values
```
Comment (biopete) [inline api/db-cdr/changelog-schema/db.changelog-18.xml]: Test this concept synonym theory out  by populating them in old cdr20180920 one that is not used and see how it works. And test query speed . in mysql

Comment (biopete) [reply api/db-cdr/changelog-schema/db.changelog-18.xml]: Also, what is the format of data going into this column? -- Ie, when we want to show synonyms how do we pull it out ?
